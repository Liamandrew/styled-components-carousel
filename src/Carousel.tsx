import React, { useState } from 'react';
import styled from 'styled-components';
import Slider from './Slider';
import Indicator from './Indicator';
import Arrow, { ArrowType } from './Arrow';
import { canGoNext, canGoPrevious, getIndexForAction, matchBreakpoint } from './helpers';

export type CarouselSettings = {
    slidesToShow?: number;
    center?: boolean;
    centerPadding?: number;
    infinite?: boolean;
    showIndicator?: boolean;
    showArrows?: boolean;
};

export type Breakpoint = {
    size: number;
    settings: CarouselSettings;
};

export type Props = CarouselSettings & {
    breakpoints?: Breakpoint[];
};

const CarouselWrapper = styled.div`
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
    position: relative;
    overflow: hidden;
    width: 100%;
`;

const SliderWrapper = styled.div`
    display: flex;
    flex-direction: row;
    align-items: center;
    width: 100%;
`;

const Component: React.FC<Props & { debug?: boolean }> = ({
    showIndicator = true,
    showArrows = true,
    slidesToShow = 1,
    centerPadding = 0,
    breakpoints,
    center,
    infinite,
    debug,
    children,
}) => {
    const childrenCount = React.Children.count(children);
    const [activeSettings, setActiveSettings] = useState(
        matchBreakpoint(
            window.innerWidth,
            {
                center,
                infinite,
                slidesToShow,
                showArrows,
                showIndicator,
                centerPadding,
            },
            breakpoints,
        ),
    );
    const [{ previousActive, active, infiniteActive }, setActive] = useState({
        previousActive: 0,
        active: 0,
        infiniteActive: 0,
    });

    return (
        <CarouselWrapper>
            <SliderWrapper>
                {activeSettings.showArrows && (
                    <Arrow
                        type={ArrowType.previous}
                        onClick={() => setActive(getIndexForAction(active, active - 1, childrenCount))}
                        disabled={!canGoPrevious(active, activeSettings.infinite)}
                    />
                )}
                <Slider
                    previousActive={previousActive}
                    active={active}
                    infiniteActive={infiniteActive}
                    onWindowResize={() =>
                        setActiveSettings(
                            matchBreakpoint(
                                window.innerWidth,
                                {
                                    center,
                                    infinite,
                                    slidesToShow,
                                    showArrows,
                                    showIndicator,
                                    centerPadding,
                                },
                                breakpoints,
                            ),
                        )
                    }
                    debug={debug}
                    {...activeSettings}
                >
                    {children}
                </Slider>
                {activeSettings.showArrows && (
                    <Arrow
                        type={ArrowType.next}
                        onClick={() => setActive(getIndexForAction(active, active + 1, childrenCount))}
                        disabled={
                            !canGoNext(
                                active,
                                childrenCount,
                                activeSettings.slidesToShow || slidesToShow,
                                activeSettings.infinite,
                            )
                        }
                    />
                )}
            </SliderWrapper>
            {activeSettings.showIndicator && (
                <Indicator
                    items={childrenCount}
                    slidesToShow={activeSettings.slidesToShow || slidesToShow}
                    infinite={activeSettings.infinite}
                    active={infiniteActive}
                    onPress={index => setActive(getIndexForAction(active, index, childrenCount))}
                />
            )}
        </CarouselWrapper>
    );
};

export default Component;
