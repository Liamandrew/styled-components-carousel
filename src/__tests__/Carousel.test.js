import React from 'react';
import { render, getByTestId } from '@testing-library/react';
import Carousel from '../Carousel';

describe('Carousel', () => {
    it('should render a carousel with default settings', () => {
        const text = 'hello world';
        const { container, getByText, getByTestId } = render(
            <Carousel>
                <div>{text}</div>
            </Carousel>,
        );

        expect(getByText(text)).toBeInTheDocument();
        expect(getByTestId('carousel-indicator-0')).toBeInTheDocument();
        expect(getByTestId('arrow-next')).toBeInTheDocument();
        expect(getByTestId('arrow-previous')).toBeInTheDocument();
        expect(container).toMatchSnapshot();
    });

    it('should not show arrows or indicators if disabled', () => {
        const text = 'hello world';
        const { queryByTestId } = render(
            <Carousel showArrows={false} showIndicator={false}>
                <div>{text}</div>
            </Carousel>,
        );

        expect(queryByTestId('carousel-indicator-0')).not.toBeInTheDocument();
        expect(queryByTestId('arrow-next')).not.toBeInTheDocument();
        expect(queryByTestId('arrow-previous')).not.toBeInTheDocument();
    });
});
