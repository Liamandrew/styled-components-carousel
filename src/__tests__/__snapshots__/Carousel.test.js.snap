// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Carousel should render a carousel with default settings 1`] = `
<div>
  <div
    class="sc-AxheI cbWmFP"
  >
    <div
      class="sc-Axmtr ljaaBB"
    >
      <button
        class="sc-AxgMl eyYDNl"
        data-testid="arrow-previous"
        disabled=""
      >
        ‹
      </button>
      <div
        class="sc-AxiKw cVFcdW"
      >
        <div
          style="cursor: pointer; transform: translate3d(0px, 0px, 0px);"
        >
          <div
            class="sc-AxjAm fdWRDs"
          >
            <div
              class="sc-AxirZ exXtW"
              width="0"
            >
              <div>
                hello world
              </div>
            </div>
          </div>
        </div>
      </div>
      <button
        class="sc-AxgMl eyYDNl"
        data-testid="arrow-next"
        disabled=""
      >
        ›
      </button>
    </div>
    <div
      class="sc-AxhCb duHzzX"
    >
      <button
        class="sc-AxhUy jiZuIL"
        data-testid="carousel-indicator-0"
      />
    </div>
  </div>
</div>
`;
